### PR TITLE
Specify RST-like behavior for CONNECTION_CLOSE. Fixes #330, #328

### LINKED ISSUE #330: CONNECTION_CLOSE semantics
From discussion on the list, we could clarify the meaning of CONNECTION_CLOSE. Specifically, it would be good to clarify if it acts more like a TCP FIN or a TCP RST. I would suggest that the analog of a TCP FIN is a FIN on a QUIC stream as they both gracefully half-close a single bidirectional communication channel. After a FIN, retransmission of data may still occur to deal with packet loss.

A QUIC CONNECTION_CLOSE, on the other hand, is an immediate termination of the connection. No more data is retransmitted. As such it is more like a TCP RST.

### PR DESCRIPTION
**Pull request by ekr**

### CODE REVIEW

**Comment by siyengar** [comment]
lgtm, maybe could use some guidance on how exactly the server could implement such a mechanism, however we can always add that later once we have more implementation experience.

**Comment by mikkelfj** [comment]
This is not entirely clear to me. If there is a requirement for backing off, but not requirement for replying to a connection close (is there?), then state will just be tied up for no good reason. If you have 100K connections per minute, and you need to hold a connection for 10 minutes to make sure you can back off, well ...

**Comment by ekr** [comment]
See my email on this topic to the list where I lay out the tradeoffs.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1469,9 +1469,17 @@ of three ways:
...
-TODO: Connections that are terminated are added to a TIME_WAIT list at the
-server, so as to absorb any straggler packets in the network.  Discuss TIME_WAIT
-list.
+After receiving either a CONNECTION_CLOSE frame or a Public Reset, an
+endpoint MUST NOT send additional packets on that connection. After
+sending either a CONNECTION_CLOSE frame or a Public Reset packet,
+implementations MUST NOT send any non-closing packets on that connection. If
+additional packets are received after this time, implementations
+SHOULD respond to them by sending either a CONNECTION_CLOSE frame or a
+Public Reset packet, either of which may just be a duplicate of a
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I recommend that you use CONNECTION_CLOSE here and instead note that an endpoint MAY send a Public Reset instead if it doesn't want to save the state necessary to generate CONNECTION_CLOSE.

There's a sliding scale here:

1. maintain full state and generate a new CONNECTION_CLOSE in a new packet
2. only keep the packet that holds CONNECTION_CLOSE and retransmit that unmodified
3. drop state and rely on Public/Stateless Reset

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -1469,9 +1469,17 @@ of three ways:
...
+previous packet. Implementations SHOULD throttle these responses, for
+instance by exponentially backing off the number of packets which must
+be received before sending a response.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: s/must be/are

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (ekr) [inline draft-ietf-quic-transport.md]: I'm happy to revise this if that's the consensus. I was trying to be flexible.

Comment (ianswett) [inline draft-ietf-quic-transport.md]: I agree that 2 and 3 are a sliding scale and that's why SHOULD makes sense to me.  I haven't come up with any reasons why option 1 is preferable to 2, so I would not suggest it, but maybe there are some reasons I haven't come up with?

Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I agree with Ian that we don't need the endpoint to maintain complete state, so 2 and 3 make sense. I'd suggest that we add a suggestion about how long a peer resends CONNECTION_CLOSE after sending it once (basically how long state should be kept in option (2)), and I propose idle_timeout as a time period that makes sense.

**Comment by janaiyengar** [comment]
lgtm. I have a non-blocking suggestion in my inline comment.

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: I think that we should allow all three, though gently discourage 3.

**Comment by ekr** [comment]
How's this look?

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: Whats the argument for allowing 1?

**Comment by janaiyengar** [comment]
LGTM. Not a major issue, but the line breaks are at odd places. If you are able to, please format them to 80 cols.

**Comment by janaiyengar** [comment]
I'd like to merge this. martinthomson@, if you think we should make the text more accommodating, let's do that separately.

**Pull request merged**